src/slack/app.rs

```diff
@@ -0,0 +1,217 @@
+    user: SlackUser,
+}
+
+fn send_response(url: &str) -> HttpResponse {
+    HttpResponse::Found()
+        .header(http::header::LOCATION, url)
+        .finish()
+}
+
+fn send_error_response() -> HttpResponse {
+    send_response("/e?identityAdded=error")
+}
+
+/**
+ * First redirect that handles getting authorization for identity scopes
+ */
+fn redirect_identity(client: web::Data<Arc<BasicClient>>, session: Session) -> impl Responder {
+    let (authorize_url, csrf_state) = client.authorize_url(CsrfToken::new_random);
+    session
+        .set("identity_csrf_state", csrf_state.secret().clone())
+        .map(|_| send_response(&authorize_url.to_string()))
+}
+
+fn auth_identity<T: AsyncCisClientTrait + 'static>(
+    cis_client: web::Data<T>,
+    user_id: UserId,
+    query: web::Query<Auth>,
+    slack_uri_data: web::Data<SlackUriData>,
+    session: Session,
+) -> Box<dyn Future<Item = HttpResponse, Error = Error>> {
+    let state = CsrfToken::new(query.state.clone());
+    let slack_token_url = format!(
+        "{}{}&code={}&team={}",
+        TOKEN_URL,
+        slack_uri_data.identity_slack_auth_params,
+        query.code.clone(),
+        slack_uri_data.team_id
+    );
+    // Check state token from im_crsf_state
+    if let Some(ref must_state) = session.get::<String>("identity_csrf_state").unwrap() {
+        if must_state != state.secret() {
+            println!("Error: Identity csrf state mismatch");
+            return Box::new(future::ok(send_error_response()));
+        }
+    } else {
+        println!("Error: Missing identity csrf state");
+        return Box::new(future::ok(send_error_response()));
+    }
+
+    let get = cis_client.clone();
+    let get_uid = user_id.user_id.clone();
+    let enforced_team_id = slack_uri_data.team_id.clone();
+    // Begin slack requests by grabbing the user_id, and access_token
+    Box::new(
+        Client::default()
+            .get(slack_token_url)
+            .header(http::header::USER_AGENT, "whoami")
+            .send()
+            .map_err(Into::into)
+            .and_then(move |mut res| res.json::<SlackUserTokenResponse>().map_err(Into::into))
+            .and_then(move |sur| {
+                if sur.team_id == enforced_team_id {
+                    Ok(sur)
+                } else {
+                    Err(format_err!("slack team_id missmatch").into())
+                }
+            })
+            .and_then(move |sur| {
+                // Now that we have the access_token, user data, and channel id, go put it in the profile
+                get.get_user_by(&get_uid, &GetBy::UserId, None)
+                    .and_then(move |profile: Profile| {
+                        update_slack(
+                            format!(
+                                "{}?channel={}&team={}",
+                                slack_uri_data.direct_message_uri, sur.user_id, sur.team_id
+                            ),
+                            sur.user.name.clone(),
```

Comment:
This returns "Leo McArdle" rather than "leo", is there no way of returning the username without getting into scope hell again?

Reply:
Not hell, but complicated. The display name would require another scope and another call to an endpoint.

Reply:
Okay, I don't think it's necessary for the moment then.